pkg/cmd/roachtest/tests/mixed_version_c2c.go

```diff
@@ -0,0 +1,442 @@
+		destMvt:             destMvt,
+		sourceStartedChan:   make(chan struct{}),
+		destStartedChan:     make(chan struct{}),
+		sp:                  sp,
+		t:                   t,
+		c:                   c,
+		fingerprintArgsChan: make(chan fingerprintArgs, 1),
+		fingerprintChan:     make(chan int64, 1),
+	}
+}
+
+type sourceTenantInfo struct {
+	name  string
+	pgurl *url.URL
+}
+
+type fingerprintArgs struct {
+	retainedTime hlc.Timestamp
+	cutoverTime  hlc.Timestamp
+}
+
+type c2cMixed struct {
+	sourceMvt, destMvt  *mixedversion.Test
+	sourceStartedChan   chan struct{}
+	destStartedChan     chan struct{}
+	sp                  replicationSpec
+	t                   test.Test
+	c                   cluster.Cluster
+	fingerprintArgsChan chan fingerprintArgs
+	fingerprintChan     chan int64
+	// midUpgradeCatchupMu attempts to prevent the source from upgrading while the
+	// destination is waiting for the stream to catch up in some mixed version
+	// state.
+	midUpgradeCatchupMu syncutil.Mutex
+
+	ingestionJobID  catpb.JobID
+	workloadStopper mixedversion.StopFunc
+}
+
+func (cm *c2cMixed) SetupHook(ctx context.Context) {
+	sourceChan := make(chan sourceTenantInfo, 1)
+
+	cm.sourceMvt.OnStartup(
+		"generate pgurl",
+		func(ctx context.Context, l *logger.Logger, r *rand.Rand, h *mixedversion.Helper) error {
+			// Enable rangefeeds, required for PCR to work.
+			l.Printf("enabling rangefeeds")
+			if err := h.System.Exec(r, "SET CLUSTER SETTING kv.rangefeed.enabled = true"); err != nil {
+				return errors.Wrap(err, "failed to enable rangefeeds")
+			}
+			close(cm.sourceStartedChan)
+
+			l.Printf("generating pgurl")
+			srcNode := cm.c.Node(1)
+			srcClusterSetting := install.MakeClusterSettings()
+			addr, err := cm.c.ExternalPGUrl(ctx, l, srcNode, roachprod.PGURLOptions{
+				VirtualClusterName: install.SystemInterfaceName,
+			})
+			if err != nil {
+				return err
+			}
+
+			pgURL, err := copyPGCertsAndMakeURL(ctx, cm.t, cm.c, srcNode, srcClusterSetting.PGUrlCertsDir, addr[0])
+			if err != nil {
+				return err
+			}
+
+			sourceChan <- sourceTenantInfo{name: h.Tenant.Descriptor.Name, pgurl: pgURL}
+
+			l.Printf("waiting for destination tenant to be created")
+			chanReadCtx(ctx, cm.destStartedChan)
+			l.Printf("done")
+
+			return nil
+		},
+	)
+
+	cm.destMvt.OnStartup("create destination tenant on standby",
+		func(ctx context.Context, l *logger.Logger, r *rand.Rand, h *mixedversion.Helper) error {
+			l.Printf("waiting to hear from source cluster")
+			sourceInfo := chanReadCtx(ctx, sourceChan)
+
+			if err := h.Exec(r, fmt.Sprintf(
+				"CREATE TENANT %q FROM REPLICATION OF %q ON '%s'",
+				destTenantName, sourceInfo.name, sourceInfo.pgurl.String(),
+			)); err != nil {
+				return errors.Wrap(err, "creating destination tenant")
+			}
+
+			if err := h.QueryRow(r, fmt.Sprintf(
+				"SELECT job_id FROM [SHOW JOBS] WHERE job_type = '%s'",
+				replicationJobType,
+			)).Scan(&cm.ingestionJobID); err != nil {
```

Comment:
The job type should be able to be a query param here rather than sprintf'ing it.